File: src/cs/PhantasyIslandPythonRemoteControl/HttpLayer.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace PhantasyIslandPythonRemoteControl
{
    public static class HttpLayer
    {
        private static readonly HttpClient _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };

        public static async Task<JsonElement?> Ping() => await SendCmd("ping");
        public static async Task<JsonElement?> PingVolatile() => await SendCmdVolatile("ping");
        public static async Task<JsonElement?> Start() => await SendCmd("start");
        public static async Task<JsonElement?> StartVolatile() => await SendCmdVolatile("start");

        public static async Task<JsonElement?> SendCmd(string s)
        {
            try
            {
                var response = await _httpClient.GetStringAsync($"http://{Config.RemoteLocation}/ECU_HTTP/sendStringCmd?c={s}");
                return JsonSerializer.Deserialize<JsonElement>(response);
            }
            catch (Exception e)
            {
                Console.WriteLine($"send_cmd {s} Error: {e.Message}");
                return JsonSerializer.Deserialize<JsonElement>("{\"ok\": false, \"r\": \"ConnectionError\"}");
            }
        }

        public static async Task<JsonElement?> SendCmdVolatile(string s)
        {
            try
            {
                var response = await _httpClient.GetStringAsync($"http://{Config.RemoteLocation}/ECU_HTTP/sendStringCmd?cc={s}");
                return JsonSerializer.Deserialize<JsonElement>(response);
            }
            catch (Exception e)
            {
                Console.WriteLine($"send_cmd_volatile {s} Error: {e.Message}");
                return JsonSerializer.Deserialize<JsonElement>("{\"ok\": false, \"r\": \"ConnectionError\"}");
            }
        }

        public static async Task<JsonElement?> GetAllAirplaneStatus()
        {
            try
            {
                var response = await _httpClient.GetStringAsync($"http://{Config.RemoteLocation}/ECU_HTTP/requestPullAllAirplaneState");
                return JsonSerializer.Deserialize<JsonElement>(response);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"ConnectionError Cannot Connect to PhantasyIsland: {e.Message}");
                throw new Exception("ConnectionError Cannot Connect to PhantasyIsland", e);
            }
        }

        public static async Task<string> GetAirplaneCameraImage(string port, string camera)
        {
            try
            {
                var response = await _httpClient.GetStringAsync($"http://{Config.RemoteLocation}/ECU_HTTP/requestPullImage?flyPort={port}&imageType={camera}");
                var json = JsonSerializer.Deserialize<JsonElement>(response);
                if (json.GetProperty("ok").GetBoolean())
                {
                    return json.GetProperty("imgDataString").GetString();
                }
                return null;
            }
            catch (Exception e)
            {
                Console.Error.WriteLine($"ConnectionError: {e.Message}");
                return null;
            }
        }

        public static Dictionary<string, object> ProcessAirplane(JsonElement j)
        {
            if (j.GetProperty("ok").GetBoolean())
            {
                var airplanes = j.GetProperty("airplanes");
                var airplaneStatus = new Dictionary<string, object>();
                foreach (var air in airplanes.EnumerateArray())
                {
                    var keyName = air.GetProperty("keyName").GetString();
                    var status = new Dictionary<string, object>
                    {
                        ["keyName"] = keyName,
                        ["typeName"] = air.GetProperty("typeName").GetString(),
                        ["updateTimestamp"] = air.GetProperty("updateTimestamp").GetDouble(),
                        ["status"] = air.GetProperty("status"),
                        ["cameraFront"] = air.GetProperty("cameraFront").TryGetProperty("imgDataString", out var frontImg) ? frontImg.GetString() : null,
                        ["cameraDown"] = air.GetProperty("cameraDown").TryGetProperty("imgDataString", out var downImg) ? downImg.GetString() : null
                    };
                    airplaneStatus[keyName] = status;
                }
                return airplaneStatus;
            }
            return null;
        }
    }
}

File: src/fs/HttpLayer.fs
namespace PhantasyIslandRemoteControl

open System
open System.Collections.Generic
open System.Net.Http
open System.Text.Json
open System.Threading.Tasks

module HttpLayer =
    let private _httpClient = new HttpClient(Timeout = TimeSpan.FromSeconds(10.0))

    let sendCmd (s: string) =
        async {
            try
                let url = sprintf "http://%s/ECU_HTTP/sendStringCmd?c=%s" Config.RemoteLocation s
                let! response = _httpClient.GetStringAsync(url) |> Async.AwaitTask
                return Some(JsonSerializer.Deserialize<JsonElement>(response))
            with e ->
                printfn "send_cmd %s Error: %s" s e.Message
                return Some(JsonSerializer.Deserialize<JsonElement>("{\"ok\": false, \"r\": \"ConnectionError\"}"))
        }

    let sendCmdVolatile (s: string) =
        async {
            try
                let url = sprintf "http://%s/ECU_HTTP/sendStringCmd?cc=%s" Config.RemoteLocation s
                let! response = _httpClient.GetStringAsync(url) |> Async.AwaitTask
                return Some(JsonSerializer.Deserialize<JsonElement>(response))
            with e ->
                printfn "send_cmd_volatile %s Error: %s" s e.Message
                return Some(JsonSerializer.Deserialize<JsonElement>("{\"ok\": false, \"r\": \"ConnectionError\"}"))
        }

    let ping () = sendCmd "ping"
    let pingVolatile () = sendCmdVolatile "ping"
    let start () = sendCmd "start"
    let startVolatile () = sendCmdVolatile "start"

    let getAllAirplaneStatus () =
        async {
            try
                let url = sprintf "http://%s/ECU_HTTP/requestPullAllAirplaneState" Config.RemoteLocation
                let! response = _httpClient.GetStringAsync(url) |> Async.AwaitTask
                return JsonSerializer.Deserialize<JsonElement>(response)
            with e ->
                eprintfn "ConnectionError Cannot Connect to PhantasyIsland: %s" e.Message
                return raise (Exception("ConnectionError Cannot Connect to PhantasyIsland", e))
        }

    let getAirplaneCameraImage (port: string) (camera: string) =
        async {
            try
                let url = sprintf "http://%s/ECU_HTTP/requestPullImage?flyPort=%s&imageType=%s" Config.RemoteLocation port camera
                let! response = _httpClient.GetStringAsync(url) |> Async.AwaitTask
                let json = JsonSerializer.Deserialize<JsonElement>(response)
                if json.GetProperty("ok").GetBoolean() then
                    return json.GetProperty("imgDataString").GetString()
                else
                    return null
            with e ->
                eprintfn "ConnectionError: %s" e.Message
                return null
        }

    let processAirplane (j: JsonElement) =
        if j.GetProperty("ok").GetBoolean() then
            let airplanes = j.GetProperty("airplanes")
            let airplaneStatus = Dictionary<string, obj>()
            for air in airplanes.EnumerateArray() do
                let keyName = air.GetProperty("keyName").GetString()
                let status = Dictionary<string, obj>()
                status.["keyName"] <- keyName
                status.["typeName"] <- air.GetProperty("typeName").GetString()
                status.["updateTimestamp"] <- air.GetProperty("updateTimestamp").GetDouble()
                status.["status"] <- air.GetProperty("status")
                
                let frontImg =
                    let mutable p = JsonElement()
                    if air.GetProperty("cameraFront").TryGetProperty("imgDataString", &p) then
                        p.GetString()
                    else
                        null
                status.["cameraFront"] <- frontImg

                let downImg =
                    let mutable p = JsonElement()
                    if air.GetProperty("cameraDown").TryGetProperty("imgDataString", &p) then
                        p.GetString()
                    else
                        null
                status.["cameraDown"] <- downImg

                airplaneStatus.[keyName] <- status
            airplaneStatus
        else
            null
